### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by ZackarySantana** [apple/pkl#128] [Q&A] [answered]
Looking in to "pkl:test", it seems to test expected output but it seems not to work with with validation. e.g. I have (before) this validation that happens that can get quite complex and I want to prevent regressions on it. I expect it to error when the configuration is set up incorrectly.

My first guess is to make my own tooling that runs it and compares the output with a file that I make. This seems really close to what the `pkl test` command does. There aren't many docs about how to test and if anyone has insight please let me know!

Ideally, my end goal is to verify that my template throws errors when I want it to and does not throw errors when I do not want it to.

```
output {
    value = new {
        ["val"] = validateVal()
    }
}
```
Note: Where validateVal is expected to either throw an error or return val

**Comment by HT154** [comment, marked as answer]
There are two main "modes" of `pkl:test`:
* `facts: Mapping<String, Listing<Boolean>>`
  * Each Listing element is asserted to be `true`, if any are false `pkl test` exits non-zero
  * There's an example of this in use here: https://github.com/apple/pkl-pantry/blob/main/packages/pkl.experimental.uri/tests/URI.pkl
* `examples: Mapping<String, Listing<unknown>>` - This entire property is evaluated to Pcf and stored in a `.pkl-expected.pcf` file on first run
  * On subsequent runs, `pkl test` validates the output remains the same
  * If it differs a `.pkl-actual.pcf` file is produced with the output and `pkl test` exits non-zero
  * Here's an example: https://github.com/apple/pkl-pantry/tree/main/packages/pkl.table/tests

You can use [`catch`](https://pkl-lang.org/package-docs/pkl/current/test/index.html#catch%28%29) with either to assert that errors are thrown (and match an expected string).
